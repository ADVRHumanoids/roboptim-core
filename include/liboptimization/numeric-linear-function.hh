// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericLinearFunction class.
 */

#ifndef OPTIMIZATION_NUMERIC_LINEAR_FUNCTION_HH
# define OPTIMIZATION_NUMERIC_LINEAR_FUNCTION_HH

# include <liboptimization/linear-function.hh>

namespace optimization
{
  /// Define a quadratic function.
  class NumericLinearFunction : public LinearFunction
  {
  public:
    NumericLinearFunction (const matrix_t&, const vector_t&) throw ();

    virtual vector_t operator () (const vector_t&) const throw ();

    virtual gradient_t gradient (const vector_t&, int) const throw ();
    virtual jacobian_t jacobian (const vector_t&) const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();

  private:
    matrix_t a_;
    vector_t b_;
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_QUADRATIC_FUNCTION_HH
