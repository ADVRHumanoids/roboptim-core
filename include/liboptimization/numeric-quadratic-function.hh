// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericQuadraticFunction class.
 */

#ifndef OPTIMIZATION_NUMERIC_QUADRATIC_FUNCTION_HH
# define OPTIMIZATION_NUMERIC_QUADRATIC_FUNCTION_HH
# include <boost/numeric/ublas/symmetric.hpp>

# include <liboptimization/quadratic-function.hh>

namespace optimization
{
  namespace ublas = boost::numeric::ublas;

  /// Define a quadratic function.
  class NumericQuadraticFunction : public QuadraticFunction
  {
  public:
    NumericQuadraticFunction (const matrix_t&, const vector_t&)
      throw ();

    virtual vector_t operator () (const vector_t&) const throw ();
    virtual gradient_t gradient (const vector_t&, int) const throw ();
    virtual hessian_t hessian (const vector_t&, int) const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  private:
    ublas::symmetric_matrix<double, ublas::lower> a_;
    vector_t b_;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_QUADRATIC_FUNCTION_HH
