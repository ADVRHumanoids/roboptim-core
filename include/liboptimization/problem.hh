// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Problem class.
 */

#ifndef OPTIMIZATION_PROBLEM_HH
# define OPTIMIZATION_PROBLEM_HH
# include <boost/optional.hpp>
# include <boost/static_assert.hpp>
# include <boost/type_traits/is_convertible.hpp>

# include <liboptimization/fwd.hh>
# include <liboptimization/function.hh>

namespace optimization
{
  /// Define a problem in a generic way.
  template <typename F, typename C>
  class Problem
  {
    BOOST_STATIC_ASSERT((boost::is_convertible<F*, Function*>::value));
  public:
    typedef F function_t;
    typedef C constraint_t;

    typedef std::vector<constraint_t> constraints_t;
    typedef boost::optional<Function::vector_t> startingPoint_t;

    explicit Problem (const function_t&) throw ();

    /// Copy constructor.
    explicit Problem (const Problem<F, C>&) throw ();

    /// Copy constructor (convert from another class of problem).
    template <typename F_, typename C_>
    explicit Problem (const Problem<F_, C_>&) throw ();

    ~Problem () throw ();

    const function_t& function () const throw ();

    const constraints_t& constraints () const throw ();
    constraints_t& constraints () throw ();

    startingPoint_t& startingPoint () throw ();
    const startingPoint_t& startingPoint () const throw ();

  private:
    const function_t& function_;
    startingPoint_t startingPoint_;
    constraints_t constraints_;
  };
}; // end of namespace optimization

# include <liboptimization/problem.hxx>
#endif //! OPTIMIZATION_PROBLEM_HH
