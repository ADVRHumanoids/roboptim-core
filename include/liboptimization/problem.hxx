// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Problem class.
 */

#ifndef OPTIMIZATION_PROBLEM_HXX
# define OPTIMIZATION_PROBLEM_HXX
# include <algorithm>
# include <boost/type_traits/remove_pointer.hpp>

namespace optimization
{
  template <typename F, typename C>
  Problem<F, C>::Problem (const function_t& f) throw ()
    : function_ (f),
      startingPoint_ (),
      constraints_ ()
  {
  }

  template <typename F, typename C>
  Problem<F, C>::~Problem () throw ()
  {
  }

  // Copy constructor.
  template <typename F, typename C>
  Problem<F, C>::Problem (const Problem<F, C>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (pb.constraints_)
  {
  }

  // Copy constructor (convert from another class of problem).
  template <typename F, typename C>
  template <typename F_, typename C_>
  Problem<F, C>::Problem (const Problem<F_, C_>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ ()
  {
    // Check that F is a subtype of F_.
    BOOST_STATIC_ASSERT((boost::is_convertible<F_*, F*>::value));
    // Check that C is a subtype of C_.

    typedef typename boost::remove_pointer<C_>::type rpC_;
    typedef typename boost::remove_pointer<C>::type rpC;
    BOOST_STATIC_ASSERT((boost::is_convertible<rpC_*, rpC*>::value));

    std::copy (pb.constraints_.begin (), pb.constraints_.end (),
               constraints_.begin ());
  }

  template <typename F, typename C>
  const typename Problem<F, C>::function_t&
  Problem<F, C>::function () const throw ()
  {
    return function_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () const throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () const throw ()
  {
    return startingPoint_;
  }

}; // end of namespace optimization

# include <liboptimization/problem.hxx>
#endif //! OPTIMIZATION_PROBLEM_HH
