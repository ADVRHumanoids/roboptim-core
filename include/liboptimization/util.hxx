// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures (templated) implementation.
 */
#ifndef OPTIMIZATION_UTIL_HXX
# define OPTIMIZATION_UTIL_HXX
# include <boost/variant/apply_visitor.hpp>

# include <liboptimization/function.hh>

namespace optimization
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }


    struct GenVariantPrintVisitor : public boost::static_visitor<void>
    {
      GenVariantPrintVisitor (std::ostream& o)
        : o_ (o)
      {}

      template <typename T>
      void operator () (const T& value)
      {
        o_ << value;
      }

      template <typename T>
      void operator () (const T* value)
      {
        if (!value)
          o_ << "Null pointer";
        else
          o_ << *value;
      }

    private:
      std::ostream& o_;
    };

  }; // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const boost::variant<T1, T2>& var)
  {
    detail::GenVariantPrintVisitor pv (o);
    boost::apply_visitor (pv, var);
    return o;
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HXX
