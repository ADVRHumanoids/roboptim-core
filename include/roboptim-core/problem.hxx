// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Problem class.
 */

#ifndef ROBOPTIM_CORE_PROBLEM_HXX
# define ROBOPTIM_CORE_PROBLEM_HXX
# include <algorithm>
# include <boost/numeric/ublas/io.hpp>
# include <boost/type_traits/is_pointer.hpp>
# include <boost/type_traits/remove_pointer.hpp>
# include <boost/variant.hpp>
# include <boost/variant/get.hpp>
# include <boost/variant/apply_visitor.hpp>

# include <roboptim-core/indent.hh>
# include <roboptim-core/util.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_problem
     @{
  */

  template <typename F, typename C>
  Problem<F, C>::Problem (const function_t& f) throw ()
    : function_ (f),
      startingPoint_ (),
      constraints_ (),
      bounds_ (),
      argBounds_ (f.n),
      scales_ (),
      argScales_ (f.n)
  {
    // Check that in the objective function m = 1 (R^n -> R).
    assert (f.m == 1);

    // Initialize bound.
    std::fill (argBounds_.begin (), argBounds_.end (), makeInfiniteBound ());
    // Initialize scale.
    std::fill (argScales_.begin (), argScales_.end (), 1.);
  }

  template <typename F, typename C>
  Problem<F, C>::~Problem () throw ()
  {
  }

  // Copy constructor.
  template <typename F, typename C>
  Problem<F, C>::Problem (const Problem<F, C>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (pb.constraints_),
      bounds_ (pb.bounds_),
      argBounds_ (pb.argBounds_),
      scales_ (pb.scales_),
      argScales_ (pb.argScales_)
  {
  }

  // Copy constructor (convert from another class of problem).
  template <typename F, typename C>
  template <typename F_, typename C_>
  Problem<F, C>::Problem (const Problem<F_, C_>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (),
      bounds_ (pb.bounds_),
      argBounds_ (pb.argBounds_),
      scales_ (pb.scales_),
      argScales_ (pb.argScales_)
  {
    // Check that F is a subtype of F_.
    BOOST_STATIC_ASSERT((boost::is_base_of<F, F_>::value));

    // Check that C is a subtype of C_.
    typedef typename boost::remove_pointer<C_>::type rpC_;
    typedef typename boost::remove_pointer<C>::type rpC;
    BOOST_STATIC_ASSERT((boost::is_base_of<rpC, rpC_>::value));

    std::copy (pb.constraints_.begin (), pb.constraints_.end (),
               constraints_.begin ());
  }

  template <typename F, typename C>
  const typename Problem<F, C>::function_t&
  Problem<F, C>::function () const throw ()
  {
    return function_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () const throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  void
  Problem<F, C>::addConstraint (const C& x, value_type s)
    throw (std::runtime_error)
  {
    addConstraint (x, makeInfiniteBound (), s);
  }

  template <typename F, typename C>
  void
  Problem<F, C>::addConstraint (const C& x, bound_t b, value_type s)
    throw (std::runtime_error)
  {
    constraints_.push_back (x);
    bounds_.push_back (b);
    scales_.push_back (s);
  }

  template <typename F, typename C>
  typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () const throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::bounds_t&
  Problem<F, C>::bounds () const throw ()
  {
    return bounds_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::bounds_t&
  Problem<F, C>::argBounds () throw ()
  {
    return argBounds_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::bounds_t&
  Problem<F, C>::argBounds () const throw ()
  {
    return argBounds_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::scales_t&
  Problem<F, C>::scales () const throw ()
  {
    return scales_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::scales_t&
  Problem<F, C>::argScales () throw ()
  {
    return argScales_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::scales_t&
  Problem<F, C>::argScales () const throw ()
  {
    return argScales_;
  }


  namespace detail
  {
    template <typename T>
    std::ostream&
    impl_print (std::ostream& o, const T* t)
    {
      return o << *t;
    }

    template <typename T>
    std::ostream&
    impl_print (std::ostream& o, const T& t)
    {
      return o << t;
    }
  }

  template <typename F, typename C>
  std::ostream&
  Problem<F, C>::print (std::ostream& o) const throw ()
  {
    o << "Problem:" << incendl;
    // Function.
    o << function () << iendl;

    // Arguments' bounds.
    o << "Argument's bounds: " << argBounds () << iendl;
    // Arguments' scales.
    o << "Argument's scales: " << argScales () << iendl;

    // Constraints.
    if (constraints ().empty ())
      o << "No constraints.";
    else
      o << "Number of constraints: " << constraints ().size ();

    typedef typename constraints_t::const_iterator citer_t;
    unsigned i = 0;
    for (citer_t it = constraints ().begin ();
         it != constraints ().end ();
         ++it, ++i)
      {
        o << iendl << incindent
          << "Constraint " << i << incindent << iendl
          << "Bounds: " << bounds ()[i] << iendl
          << "Scales: " << scales ()[i] << iendl
          << decindent << decindent;
      }

    // Starting point.
    if (startingPoint_)
      o << iendl << "Starting point: " << *startingPoint_;
    else
      o << iendl << "No starting point.";

    // Infinity.
    o << iendl << "Infinity value (for all functions): " << Function::infinity ();

    // Arguments' bounds.
    // Arguments' scales.


    return o << decindent;
  }

  template <typename F, typename C>
  std::ostream&
  operator<< (std::ostream& o, const Problem<F, C>& pb)
  {
    return pb.print (o);
  }
  /**
     @}
  */
}; // end of namespace roboptim
# include <roboptim-core/problem.hxx>
#endif //! ROBOPTIM_CORE_PROBLEM_HH
