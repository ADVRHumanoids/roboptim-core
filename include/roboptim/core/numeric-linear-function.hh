// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_NUMERIC_LINEAR_FUNCTION_HH
# define ROBOPTIM_CORE_NUMERIC_LINEAR_FUNCTION_HH

# include <roboptim/core/linear-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /// \brief Build a linear function from a vector and a matrix.
  ///
  /// Implement a linear function using the general formula:
  /// \f[f(x) = A x + b\f]
  /// where \f$A\f$ and \f$b\f$ are set when the class is instantiated.
  class NumericLinearFunction : public LinearFunction
  {
  public:
    /// \brief Build a linear function from a matrix and a vector.
    ///
    /// See class documentation for A and b definition.
    /// \param A A matrix
    /// \param b b vector
    NumericLinearFunction (const matrix_t& A, const vector_t& b) throw ();
    ~NumericLinearFunction () throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, int = 0) const throw ();
    void impl_jacobian (jacobian_t&, const vector_t&) const throw ();

  private:
    /// \brief A matrix.
    matrix_t a_;
    /// \brief B vector.
    vector_t b_;
  };

  /// Example shows numeric linear function use.
  /// \example numeric-linear-function.cc

  /// @}

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_QUADRATIC_FUNCTION_HH
