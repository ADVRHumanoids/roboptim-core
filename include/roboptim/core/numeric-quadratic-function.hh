// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericQuadraticFunction class.
 */

#ifndef ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HH
# define ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HH
# include <boost/numeric/ublas/symmetric.hpp>

# include <roboptim/core/quadratic-function.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_function
     @{
  */
  namespace ublas = boost::numeric::ublas;

  /// Define a quadratic function.
  class NumericQuadraticFunction : public QuadraticFunction
  {
  public:
    NumericQuadraticFunction (const matrix_t&, const vector_t&)
      throw ();

    ~NumericQuadraticFunction () throw ();

    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, int = 0) const throw ();
    void impl_hessian (hessian_t& hessian,
		       const argument_t& argument,
		       int functionId = 0) const throw ();
  private:
    ublas::symmetric_matrix<double, ublas::lower> a_;
    vector_t b_;
  };
/**
   @}
*/

} // end of namespace roboptim
#endif //! ROBOPTIM_CORE_QUADRATIC_FUNCTION_HH
