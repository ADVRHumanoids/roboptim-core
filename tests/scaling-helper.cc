// Copyright (C) 2016 by Benjamin Chr√©tien, CNRS-AIST JRL.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/problem.hh>
#include <roboptim/core/scaling-helper.hh>
#include <roboptim/core/function/constant.hh>
#include <roboptim/core/numeric-linear-function.hh>
#include <roboptim/core/numeric-quadratic-function.hh>

using namespace roboptim;

typedef boost::mpl::list<EigenMatrixDense> functionTypes_t;

// Output stream
boost::shared_ptr<boost::test_tools::output_test_stream> output;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE_TEMPLATE (scaling_helper, T, functionTypes_t)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("scaling-helper");

  typedef Problem<T> problem_t;
  typedef typename problem_t::function_t      function_t;
  typedef typename function_t::argument_t     argument_t;
  typedef typename problem_t::intervals_t     intervals_t;
  typedef typename problem_t::interval_t      interval_t;
  typedef typename problem_t::scaling_t       scaling_t;

  typedef GenericConstantFunction<T>          constantFunction_t;
  typedef GenericNumericLinearFunction<T>     numericLinearFunction_t;
  typedef GenericNumericQuadraticFunction<T>  numericQuadraticFunction_t;

  // Initialize problem
  const size_t n = 2;
  typename constantFunction_t::vector_t v (n);
  v << 1., 2.;
  boost::shared_ptr<constantFunction_t>
    f = boost::make_shared<constantFunction_t> (v);
  (*output) << *f << std::endl;

  problem_t pb (f);
  (*output) << pb << std::endl;

  std::vector<argument_t> x (2, argument_t::Zero (n));
  x[0] << 2., 1.;
  x[1] << 100., -200.;

  (*output) << x[0] << std::endl;
  (*output) << x[1] << std::endl;

  // Check without constraints
  {
    ScalingHelper<T> sh (pb, function_t::makeInterval (1e-1, 1e2));
    (*output) << sh << std::endl;
  }

  // Add some dummy constraints
  typename numericLinearFunction_t::matrix_t a (2,2);
  a.setZero ();
  a.coeffRef (0,0) = 300.;
  a.coeffRef (1,1) = 10.;
  typename numericLinearFunction_t::vector_t b (2);
  b << 200., 300.;
  boost::shared_ptr<numericLinearFunction_t>
    g0 = boost::make_shared<numericLinearFunction_t> (a, b);
  intervals_t intervals (2);
  scaling_t scaling (2, 1);
  for (size_t i = 0; i < intervals.size (); ++i)
    intervals[i] = Function::makeLowerInterval (0.);
  pb.addConstraint (g0, intervals, scaling);

  boost::shared_ptr<numericQuadraticFunction_t>
    g1 = boost::make_shared<numericQuadraticFunction_t> (a, b);
  intervals.resize (1);
  scaling.resize (1);
  for (size_t i = 0; i < intervals.size (); ++i)
    intervals[i] = Function::makeLowerInterval (0.);
  pb.addConstraint (g1, intervals, scaling);

  a *= 1e-6;
  b *= 1e-6;
  boost::shared_ptr<numericQuadraticFunction_t>
    g2 = boost::make_shared<numericQuadraticFunction_t> (a, b);
  intervals.resize (1);
  scaling.resize (1);
  for (size_t i = 0; i < intervals.size (); ++i)
    intervals[i] = Function::makeLowerInterval (0.);
  pb.addConstraint (g2, intervals, scaling);

  (*output) << pb << std::endl;

  // Check with constraints
  {
    ScalingHelper<T> sh (pb, function_t::makeInterval (1e-1, 1e2));
    (*output) << Eigen::MatrixXd (pb.jacobian (x[0])) << std::endl;
    (*output) << Eigen::MatrixXd (pb.jacobian (x[1])) << std::endl;
    (*output) << sh << std::endl;
    sh.computeScaling (x);
    (*output) << sh << std::endl;
  }

  {
    interval_t bounds;
    bounds.first = -1.;
    bounds.second = 100.;
    BOOST_CHECK_THROW (ScalingHelper<T> sh1 (pb, bounds),
          std::runtime_error);
    bounds.first = 200.;
    BOOST_CHECK_THROW (ScalingHelper<T> sh2 (pb, bounds),
          std::runtime_error);
  }

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
