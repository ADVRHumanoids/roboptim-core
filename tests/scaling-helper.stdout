Constant function
  offset = [2](1,2)
Problem:
  Constant function
    offset = [2](1,2)
  Objective scaling: 1, 1
  Arguments bounds: (-inf, inf), (-inf, inf)
  Arguments scaling: 1, 1
  No constraints.
  No starting point.
  Infinity value (for all functions): inf
[2](2,1)
[2](100,-200)
No suggested scaling parameters
Problem:
  Constant function
    offset = [2](1,2)
  Objective scaling: 1, 1
  Arguments bounds: (-inf, inf), (-inf, inf)
  Arguments scaling: 1, 1
  Number of constraints: 3
  Constraint 0
      Numeric linear function:
        A = [2,2]((300,0), (0,10))
        B = [2](200,300)
      Bounds: (0, inf), (0, inf)
      Scaling: 1, 1
  Constraint 1
      Numeric quadratic function:
        A = [2,2]((300,0), (0,10))
        B = [2](200,300)
        c = [1](0)
      Bounds: (0, inf)
      Scaling: 1
  Constraint 2
      Numeric quadratic function:
        A = [2,2]((0.0003,0), (0,1e-05))
        B = [2](0.0002,0.0003)
        c = [1](0)
      Bounds: (0, inf)
      Scaling: 1
  No starting point.
  Infinity value (for all functions): inf
[4,2]((300,0), (0,10), (1400,320), (0.0014,0.00032))
[4,2]((300,0), (0,10), (60200,-3700), (0.0602,-0.0037))
Suggested scaling parameters:
  Argument scaling: 1, 1
  Constraint 0:
    Numeric linear function:
      A = [2,2]((300,0), (0,10))
      B = [2](200,300)
    Scaling: 1, 1
  Constraint 1:
    Numeric quadratic function:
      A = [2,2]((300,0), (0,10))
      B = [2](200,300)
      c = [1](0)
    Scaling: 1
  Constraint 2:
    Numeric quadratic function:
      A = [2,2]((0.0003,0), (0,1e-05))
      B = [2](0.0002,0.0003)
      c = [1](0)
    Scaling: 1
Suggested scaling parameters:
  Argument scaling: 1, 1
  Constraint 0:
    Numeric linear function:
      A = [2,2]((300,0), (0,10))
      B = [2](200,300)
    Scaling: 0.333333, 1
  Constraint 1:
    Numeric quadratic function:
      A = [2,2]((300,0), (0,10))
      B = [2](200,300)
      c = [1](0)
    Scaling: 0.00166113
  Constraint 2:
    Numeric quadratic function:
      A = [2,2]((0.0003,0), (0,1e-05))
      B = [2](0.0002,0.0003)
      c = [1](0)
    Scaling: 71.4286

